flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/table/KinesisDynamicTableFactoryTest.java

```diff
@@ -0,0 +1,358 @@
+import static org.apache.flink.streaming.connectors.kinesis.table.RowDataKinesisDeserializationSchema.Metadata;
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Test for {@link KinesisDynamicSource} and {@link KinesisDynamicSink} created
+ * by {@link KinesisDynamicTableFactory}.
+ */
+public class KinesisDynamicTableFactoryTest extends TestLogger {
+
+	private static final String SOURCE_STREAM = "sourceStream";
+	private static final String SINK_STREAM = "targetStream";
+
+	private static final String SOURCE_TABLE = "sourceTable";
+	private static final String SINK_TABLE = "sinkTable";
+
+	private static final Properties KINESIS_PROPERTIES = new Properties() {{
+		setProperty("aws.region", "us-west-2");
+	}};
+
+	@Rule
+	public ExpectedException thrown = ExpectedException.none();
+
+	// --------------------------------------------------------------------------------------------
+	// Positive tests
+	// --------------------------------------------------------------------------------------------
+
+	@Test
+	public void testGoodTableSource() {
+		TableSchema sourceSchema = defaultSourceSchema().build();
+		Map<String, String> sourceOptions = defaultSourceOptions().build();
+
+		// Construct actual DynamicTableSource using FactoryUtil
+		KinesisDynamicSource actualSource = (KinesisDynamicSource)
+			FactoryUtil.createTableSource(
+				null,
+				ObjectIdentifier.of("default", "default", SOURCE_TABLE),
+				createSourceTable(sourceSchema, sourceOptions, Collections.emptyList()),
+				new Configuration(),
+				Thread.currentThread().getContextClassLoader(),
+				false);
+
+		// Construct expected DynamicTableSink using factory under test
+		KinesisDynamicSource expectedSource = new KinesisDynamicSource(
+			sourceSchema.toPhysicalRowDataType(),
+			SOURCE_STREAM,
+			KINESIS_PROPERTIES,
+			new TestFormatFactory.DecodingFormatMock(",", true));
+
+		// verify that the constructed DynamicTableSink is as expected
+		assertEquals(expectedSource, actualSource);
+
+		// verify that the copy of the constructed DynamicTableSink is as expected
+		assertEquals(expectedSource, actualSource.copy());
+
+		// verify produced sink
+		ScanTableSource.ScanRuntimeProvider functionProvider =
+			actualSource.getScanRuntimeProvider(ScanRuntimeProviderContext.INSTANCE);
+		SourceFunction<RowData> sourceFunction =
+			as(functionProvider, SourceFunctionProvider.class).createSourceFunction();
+		assertThat(sourceFunction, instanceOf(FlinkKinesisConsumer.class));
+	}
+
+	@Test
+	public void testGoodTableSourceWithMetadataFields() {
+		TableSchema sourceSchema = defaultSourceSchema().build();
+		Map<String, String> sourceOptions = defaultSourceOptions().build();
+
+		// Construct actual DynamicTableSource using FactoryUtil
+		KinesisDynamicSource actualSource = (KinesisDynamicSource)
+			FactoryUtil.createTableSource(
```

Comment:
nit: move this call into a helper method also for the sink

Reply:
Done.